### ISSUE TITLE
Jobs run half as often as they should

### ISSUE THREAD

**Issue by danarnold** [ondrejbartas/sidekiq-cron#89]
After upgrading sidekiq to > 4.0 and sidekiq-cron to > 0.3.1, all of our scheduled jobs executed at exactly or almost exactly half of the frequency that they should have. For example, a job that runs once per minute would simply run once every two minutes. I'm not sure what the cause of this is, but reverting back to 0.3.1 with sidekiq 3.5.4 fixed it.

**Comment by sqbell** [comment]
We also are missing jobs after upgrade.

**Comment by mbhnyc** [comment]
I'm seeing jobs that are set to run ones a day, like `0 15 * * *` running once an hour.

**Comment by jclusso** [comment]
Any idea on this @ondrejbartas? It's stopping us from upgrading to Sidekiq 4 which has some really big improvements.

**Comment by jclusso** [comment]
Any idea on this @ondrejbartas? It's stopping us from upgrading to Sidekiq 4 which has some really big improvements.

**Comment by jclusso** [comment]
Any idea on this @ondrejbartas? It's stopping us from upgrading to Sidekiq 4 which has some really big improvements.

**Comment by jclusso** [comment]
@ondrejbartas we are still eagerly waiting a response for almost a month on this :(

**Comment by gdeglin** [comment]
Same issue here

**Comment by phyzalis** [comment]
Anything new about that?

**Comment by jclusso** [comment]
@phyzalis I gave up waiting and was looking at https://github.com/moove-it/sidekiq-scheduler but this seems to run 2x of what it should so we're just currently stuck using Sidekiq 3 still :(

**Comment by phyzalis** [comment]
I will have a look thank you. I need to be able to add jobs programmatically and this sidekiq-cron was perfect. I will have a look inside sidekiq-cron source too when I'll have time.
Thanks

**Comment by phyzalis** [comment]
For me the issue comes from RufusScheduler here:

     Rufus::Scheduler::CronLine.new(@cron).previous_time(now)

https://github.com/ondrejbartas/sidekiq-cron/blob/master/lib/sidekiq/cron/job.rb#L443

I will continue to investigate

**Comment by danarnold** [comment]
@phyzalis Can you elaborate on that? I see that RufusScheduler supports extensions to the standard cron format, but it also claims to support standard cron format.

**Comment by phyzalis** [comment]
In fact my issue was my cron format, I was having 6 elements in it instead of 5.
Fixing this it works perfectly

For example to run something every day at 10, I was using:
0 10 * * * *
Which is wrong and run my task every hour at 10 minutes

The right way was : 0 10 * * *

But I don't think this will help you

**Comment by danarnold** [comment]
Yeah, we were using standard 5 element cron.

**Comment by ondrejbartas** [comment]
@danarnold Hi sorry for delay. How can I help you? I am not really sure what is your problem.

**Comment by danarnold** [comment]
That's fine. I'm not sure where you're confused, I described the problem in the my first comment on the issue. Is there anything I can clarify? Essentially, jobs are running 1/2 as often as they should.

**Comment by hderms** [comment]
@danarnold Can you give us the exact versions of the gems you are running? I created a dirt simple gem which just scheduled a job to run every minute and reports the amount of time between each invocation but can't reproduce the issue. Also it would be helpful to get your sidekiq/sidekiq-cron config. 

I wrote this example worker, which although poorly written, provides a reasonable approximation to how often the cron task is running. https://gist.github.com/hderms/bf484c103573f65942f78cff5243d845

**Comment by danarnold** [comment]
This was with sidekiq 4.0.1, sidekiq-cron 0.4.2.

Here's our cron file with the worker names changed (they are all unique in our app, of course):

```
redacted_worker:
  cron: '*/10 * * * *'
  class: 'RedactedWorker'
  queue: low

redacted_worker:
  cron: '*/1 * * * *'
  class: 'RedactedWorker'
  queue: slow

redacted_worker:
  cron: '*/1 * * * *'
  class: 'RedactedWorker'
  queue: slow

redacted_worker:
  cron: '*/1 * * * *'
  class: 'RedactedWorker'
  queue: slow

redacted_worker:
  cron: '*/5 * * * *'
  class: 'RedactedWorker'
  queue: low

redacted_worker:
  cron: '0 * * * *'
  class: 'RedactedWorker'
  queue: slow

redacted_worker:
  cron: '0 0 */1 * *'
  class: 'RedactedWorker'
  queue: slow

redacted_worker:
  cron: '0 */2 * * *'
  class: 'RedactedWorker'
  queue:  slow

redacted_worker:
  cron: '0 */1 * * *'
  class: 'RedactedWorker'
  queue: low

redacted_worker:
  cron: '*/5 * * * *'
  class: 'RedactedWorker'
  queue: low

redacted_worker:
  cron: '*/5 * * * *'
  class: 'RedactedWorker'
  queue: low
```

It's worth mentioning that we run these workers in production on Heroku with multiple sidekiq instances, so Heroku fires up at least one dyno per queue, meaning that you get dyno processes running `sidekiq -q slow`, running `sidekiq -q low`, etc. This doesn't cause issues using the version we're using right now (sidekiq-cron 0.3.1, sidekiq 3.5.4).

**Comment by hderms** [comment]
Thanks @danarnold I will see if I can craft a minimal example project which reproduces the issue.

**Comment by hderms** [comment]
@danarnold can you reproduce this issue on your development machine or has it only been encountered on heroku?

**Comment by danarnold** [comment]
I didn't try, because I had it happening on Heroku. If need be I could try to reproduce it on my machine.

**Comment by jclusso** [comment]
@hderms, @danarnold and I work together and I can confirm it happens in both our staging and my local environment.

**Comment by hderms** [comment]
@jclusso good to know. I tried crafting a repo with the bare minimum setup
(sidekiq-cron,  multiple cron workers registered in YAML file and running
multiple sidekiq processes) but can't reproduce the issue. Are there any
other quirks of the environment you think might be important offhand?

**Comment by jclusso** [comment]
@hderms I can't think of any quirks. I'd love to hear @gdeglin chime in since he reported the same problem. I've also attached our sidekiq config.

```ruby
host = ENV['redis_host']
port = ENV['redis_port']
db = ENV['redis_db_sidekiq']
namespace = ENV['redis_namespace']

if Rails.env.development? || Rails.env.test?
  redis_url = "redis://#{host}:#{port}/#{db}"
else
  password = ENV['redis_password']
  redis_url = "redis://:#{password}@#{host}:#{port}/#{db}"
end

Sidekiq.configure_server do |config|
  config.redis = { url: redis_url, namespace: namespace }
  schedule_file = 'config/schedule.yml'

  if File.exist?(schedule_file)
    Sidekiq::Cron::Job.load_from_hash YAML.load_file(schedule_file)[Rails.env]
  end
end

Sidekiq.configure_client do |config|
  config.redis = { url: redis_url, namespace: namespace }
end
```

**Comment by gdeglin** [comment]
Nothing unusual here either. We have a task that should run every 5 minutes, and once ever 4 hours or so it would skip over one of the scheduled times. My only theory is that it could be NTPd related, but I haven't had a chance to try disabling NTPd to see if it fixes it.

**Comment by hderms** [comment]
@jclusso is @gdeglin sharing info which is consistent with how the issue presents in your experience i.e. once every X amount of time there is a deviation or is it presenting differently. I'm concerned about this issue but can't figure out how to reproduce it. My limited example app is available at: https://gist.github.com/hderms/b769dc573394db61d52531b464de4980

**Comment by jclusso** [comment]
@hderms we will need to to do further analysis on the pattern to see if that is the case. From what we see it appears to just be doing half the number of workers.

**Comment by danarnold** [comment]
I don't believe it correlates with that pattern at all. It is a simple 1/2 jobs.

**Comment by ondrejbartas** [comment]
Have you tried to setup: `average_scheduled_poll_interval` or `poll_interval` in sidekiq configuration?

**Comment by danarnold** [comment]
@ondrejbartas I'm not sure what that would do. We have it set to the sidekiq default, which is approximately every 15 seconds balanced over the number of workers that are running. I see no reason to change it.

**Comment by hderms** [comment]
@ondrejbartas we upgraded sidekiq at my current workplace and our jobs are running at the specified frequency. The evidence right now, both my experience with my work's production codebase and my stripped down sample sidekiq project provided earlier in this thread, seems to suggest there is something unusual going on that we can't seem to capture without more information. If you guys can reproduce a minimal project which has jobs running at half frequency that would be extremely helpful in debugging.

**Comment by EvanMcEwen** [comment]
I'm having a possible related issue? My jobs are running mostly fine but every so often they will simply not be added to the queue. Most of my jobs run once a day and I've noticed about twice this month they simply did not run and the sidekiq log shows no errors (it wasn't even added!). I've since added a job that runs every minute and every so often I can see that it simply skips a minute and doesn't even run. Going through my own logs and audits it looks like this issue present itself around March of 2016 for me which is precisely when I updated to Sidekiq (4.1.1) and Sidekiq-cron (0.4.2)

Before this there is no evidence of any issues and I was running Sidekiq (3.3.4) and Sidekiq-cron (0.3.1).

I'm going to revert one of my clients back to these versions and see if the issue disappears.

**Comment by jclusso** [comment]
By the way our issues ended up being long running workers getting shutdown before completion by our autoscaling of Heroku dynos. We ended up implementing locking mechanisms. We also stopped autoscaling worker queues.

**Comment by EvanMcEwen** [comment]
Just realized I never came back with what happened.

Reverting Sidekiq and Sidekiq-cron back fixed my issues. Not sure if both or only one of them was causing the problem but I'm not willing to test this in production at the moment since everything is now back to normal.

**Comment by sztosz** [comment]
Hi i had a similar problem. Using default sidekiq config `poll_interval_average: nil, average_scheduled_poll_interval: 15` which with one running process (just for the cron) it defaults to `process_count *  average_scheduled_poll_interval = 15` seconds. Yet I observed that sidekiq-cron is checking `cron-table` every 60+ seconds, like minute and 5s, or minute and 10s. 

And that's really strange, because that app's only job is to clear some keys from redis every day at midnight, and provide web gui for other apps using sidekiq and this redis instance. So basically most of the time it sits there idle.

I don't know why this happens, wasn't able to track it down. But since the execution time window for job to be started according to cron is time from cron-table + 60s and with pooling interval a slightly more than 60s you can sometimes miss the job execution. My simple naive solution was simply removing that 60s execution window https://github.com/sztosz/sidekiq-cron/commit/9d2eb40fa674e8ec4e22720499275157647b040d

@ondrejbartas I was curious if you can help with tracking what in sidekiq-cron can make checking cron-table so rarely (60+ seconds) instead of default 15 sec of pooling interval. Or maybe if there is a workaround other than me deleting that one particular line (link in paragraph above)?

**Comment by tmartyny** [comment]
@ondrejbartas We are running sidekiq pro 4.1.1 and sidekiq-cron 0.4.3 and our scheduled sidekiq-cron jobs are randomly not being enqueued. An important and time sensitive job didn't run on Sunday which made us aware of the situation, but after further research, there are more jobs that are not enqueueing consistently. Is there any progress on this issue?  The changelog looks like it is not up to date, so I don't know what 0.4.1, 0.4.2, and 0.4.3 changed. 
Thanks!
Teresa

**Comment by ondrejbartas** [comment]
#153 #150 #152

**Comment by ondrejbartas** [comment]
released 0.6.0
